userspace/Build.mk

```diff
@@ -0,0 +1,92 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+BUILD_SUBDIRS := $(addprefix userspace/,dcrypto_test gpio_test h1b_tests \
+                   sha_test spin u2f_app u2f_test)
+
+.PHONY: userspace/build
+userspace/build: $(addsuffix /build,$(BUILD_SUBDIRS))
+
+.PHONY: userspace/check
+userspace/check:
+	cd userspace && cargo check --release
+
+.PHONY: userspace/clean
+userspace/clean:
+	rm -f userspace/Cargo.lock
+
+.PHONY: userspace/devicetests
+userspace/devicetests: $(addsuffix /devicetests,$(BUILD_SUBDIRS))
+
+.PHONY: userspace/doc
+userspace/doc:
+	cd userspace && cargo doc --release
+
+.PHONY: userspace/localtests
+userspace/localtests: $(addsuffix /localtests,$(BUILD_SUBDIRS))
+
+include $(addsuffix /Build.mk,$(BUILD_SUBDIRS))
+
+# ------------------------------------------------------------------------------
+# Build rules shared between the C apps.
+# ------------------------------------------------------------------------------
```

Comment:
A comment here explaining what this design pattern of foreach statement is doing would be useful.